File: 556-Next Greater Element III.py
class Solution(object):
    def nextGreaterElement(self, n):
        """
        :type n: int
        :rtype: int
        """
        arr = list(str(n))
        if len(arr) <= 1:
            return -1
        
        for i in range(len(arr)-2, -1, -1):
            if int(arr[i]) < int(arr[i+1]):
                index = len(arr)-1
                while arr[index] <= arr[i]:
                    index-=1
                arr[i], arr[index] = arr[index], arr[i]
                arr[i+1:] = arr[i+1:][::-1]
                break
            if i == 0:
                return -1
            
        res = int("".join(arr))
        if res > 2147483647:
            return -1
        else:
            return res

File: JavaScript/556-Next Greater Element III.js
/**
 * @param {number} n
 * @return {number}
 */
var nextGreaterElement = function(n) {
    
    let arr = n.toString().split("");
    
    if (arr.length === 1) return -1;
    
    for (let i= arr.length-2; i>=0; i--){
        if (arr[i] < arr[i+1]){
            let index = arr.length-1;
            while (arr[i] >= arr[index]){
                index--;
            }
            [arr[i], arr[index]] = [arr[index], arr[i]];
            
            let left = i+1, right = arr.length-1;
            while(left<right){
                [arr[left], arr[right]] = [arr[right], arr[left]];
                left++; right--;
            }
            break;
        }
        if (i===0){
            return -1;
        }
    }
    
    let res = parseInt(arr.join(""));
    if (res > 2147483647){
        return -1;
    }else{
        return res;
    }
};